keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/vetos/RequiredSecurityGroupVeto.kt

```diff
@@ -0,0 +1,85 @@
+package com.netflix.spinnaker.keel.ec2.vetos
+
+import com.netflix.spinnaker.keel.api.Resource
+import com.netflix.spinnaker.keel.api.ec2.ClassicLoadBalancerSpec
+import com.netflix.spinnaker.keel.clouddriver.CloudDriverCache
+import com.netflix.spinnaker.keel.clouddriver.CloudDriverService
+import com.netflix.spinnaker.keel.core.api.DEFAULT_SERVICE_ACCOUNT
+import com.netflix.spinnaker.keel.ec2.CLOUD_PROVIDER
+import com.netflix.spinnaker.keel.veto.Veto
+import com.netflix.spinnaker.keel.veto.VetoResponse
+import kotlinx.coroutines.runBlocking
+import org.slf4j.LoggerFactory
+import org.springframework.stereotype.Component
+import retrofit2.HttpException
+
+@Component
+class RequiredSecurityGroupVeto(
+  private val cloudDriver: CloudDriverService,
+  private val cloudDriverCache: CloudDriverCache
+) : Veto {
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+
+  override fun check(resource: Resource<*>): VetoResponse {
+    val allow = when (val spec = resource.spec) {
+      is ClassicLoadBalancerSpec -> {
+        spec.dependencies.securityGroupNames.all { securityGroupName ->
+          spec.locations.regions.map { it.name }.all { region ->
+            runBlocking { // TODO: proper coroutine scope
+              securityGroupExists(
+                spec.locations.account,
+                spec.locations.subnet,
+                region,
+                securityGroupName
+              )
+            }
+          }
+        }
+      }
+      else -> true
+    }
+    return VetoResponse(allow, name())
+  }
+
+  private suspend fun securityGroupExists(
+    account: String,
+    vpcName: String?,
+    region: String,
+    name: String
+  ): Boolean {
+    return runCatching {
+      cloudDriver.getSecurityGroup(
+        DEFAULT_SERVICE_ACCOUNT,
+        account,
+        CLOUD_PROVIDER,
+        name,
+        region,
+        if (vpcName == null) null else cloudDriverCache.networkBy(vpcName, account, region).id
+      )
+    }
+      .onFailure { ex ->
+        if (ex is HttpException && ex.code() == 404) {
+          log.warn("security group $name not found in $account/$vpcName/$region")
+        } else {
+          log.error("error finding security group $name in $account/$vpcName/$region", ex)
+        }
+      }
+      .isSuccess
+  }
+
+  override fun messageFormat(): Map<String, Any> {
+    TODO("not implemented")
+  }
+
+  override fun passMessage(message: Map<String, Any>) {
+    TODO("not implemented")
+  }
+
+  override fun currentRejections(): List<String> {
+    TODO("not implemented")
+  }
+
+  override fun currentRejectionsByApp(application: String): List<String> {
+    TODO("not implemented")
+  }
```

Comment:
Can anyone help me understand what these methods are supposed to do?

Reply:
So, the messages are a generic way to tell a veto to do something. The veto tells you what data it accepts (`messageFormat()`), and you send it that data via `passMessage(...)`. That avoids having to make a new controller for each veto, and could let us do remote calls to vetos. For example, that's how I first implemented pausing an application, and resuming an application.

`currentRejections` is a list of things the veto is rejecting (so you don't have to look at the database). Like, when there was a veto for pausing applications, this would return a list of the applications. `currentRejectionsByApp` returns a list of things the veto is rejecting by app like if the veto was rejecting resources. 

Maybe a lot has changed since I wrote this generic stuff. The idea was to make it easy to add new vetos without having to do a lot of wiring through keel and gate. However, that might have not been worth it.

As it is, you don't have to implement any of those things for the veto to veto stuff. You just won't be able to interact with it.

Reply:
Okay, it sounds like it would make sense for me to implement `currentRejectionsByApp` but I probably don't need any of the others. Does that make sense?

Reply:
I think so. If you don't want the user to interact with the veto (for example, be able to say "i know best please actually check this resource") then that seems right.

Reply:
Yeah, I think in this case it doesn't make sense for a user to be able to say "no, I'm 100% sure that security group exists"!

Reply:
I still think this API seems kinda weird in that each veto has to track its own rejections rather than having that centralized. I don't want to tackle any changes to it here since I'd like to keep this PR focused in scope, and I'm not yet sure exactly how I'd flip that around.

Reply:
Let's talk about this in person, I think we can definitely make this better